### PR TITLE
Migrated loss functions to ModuleInfos

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)

#### On test times
This PR increases test time by ~58s locally 
Before this PR:
```
>>> python test/test_nn.py -k Loss
Ran 1003 tests in 28.977s
```
After this PR 
```
>>> python test/test_nn.py -k Loss
Ran 368 tests in 23.073s
```

```
>>> python test/test_modules.py -k Loss
Ran 836 tests in 63.900s
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115584
* #115617

### CODE REVIEW

File: torch/testing/_internal/common_nn.py [hunk 1]
```diff
@@ -2953,7 +2944,7 @@ def _multimarginloss_reference(input, target_idx, p, margin, weight):
     output = 0
     for i in range(0, len(input)):
         if i != target_idx:
-            output += max(0, weight[target_idx] * (margin - input[target_idx] + input[i]) ** p)
+            output += weight[target_idx] * (max(0, (margin - input[target_idx] + input[i])) ** p)
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_nn.py:2947]: This reference function was implemented wrongly, see docs
Screenshot 2023-12-11 at 3 38 01 PM
and the implementation

https://github.com/pytorch/pytorch/blob/585aea6e770f601159e85fa4cc39ce7f67155fdf/aten/src/ATen/native/LossMultiMargin.cpp#L37-L45

File: test/test_modules.py [hunk 2]
```diff
@@ -304,6 +304,8 @@ def _traverse_obj(self, obj, func):
             return {name: self._traverse_obj(o, func) for name, o in obj.items()}
         elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)):
             return func(obj)
+        elif isinstance(obj, (int, float, bool, complex, str)):
```
Comment (mikaylagawarecki) [inline test/test_modules.py:307]: Changed this because otherwise it was converting `Tuple[int]` to `Tuple[None]` in the noncontiguous outputs when converting lists of integers in `test_non_contiguous_tensors`

File: torch/testing/_internal/common_modules.py [hunk 3]
```diff
@@ -1657,6 +2160,158 @@ def module_inputs_torch_nn_LogSigmoid(module_info, device, dtype, requires_grad,
...
+        ('reduction_none', {'reduction': 'none'}),
+    ]
+
+    module_inputs = []
+    for desc, constructor_kwargs in cases:
+        def reference_fn(m, p, i, t, constructor_kwargs=constructor_kwargs):
+            return multilabelmarginloss_reference(i, t, **constructor_kwargs)
+
+        module_inputs.append(
+            ModuleInput(constructor_input=FunctionInput(**constructor_kwargs),
```
Comment (janeyx99) [inline torch/testing/_internal/common_modules.py:2208]: There's another MultilabelMarginLoss for 2d where the inputsize is (5, 10). Is that one omitted here cuz redundancy?

File: torch/testing/_internal/common_modules.py [hunk 4]
```diff
@@ -3103,11 +3831,60 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
...
+               skips=(
+                   # No channels_last support for loss functions.
+                   DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_memory_format'),
+                   DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),)
+               ),
+    ModuleInfo(torch.nn.BCELoss,
+               module_inputs_func=module_inputs_torch_nn_BCELoss,
+               skips=(
+                   # No channels_last support for loss functions.
+                   DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_memory_format'),
```
Comment (janeyx99) [inline torch/testing/_internal/common_modules.py:3858]: Is there a way to defacto have these be a part of all loss functions? Like the default skip list has one item already if it's a loss function, and the one item is `DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_memory_format')`?

File: torch/testing/_internal/common_nn.py [hunk 5]
```diff
@@ -3078,918 +3069,7 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
...
-        reference_fn=lambda i, t, il, tl, m:
-            ctcloss_reference(i, t, il, tl, blank=0, reduction=get_reduction(m)),
-        check_forward_only=True,
-        check_sum_reduction=True,
-        check_gradgrad=False,
-        check_half=False,
-        default_dtype=torch.double,
-    ),
-]
+criterion_tests = []
```
Comment (janeyx99) [inline torch/testing/_internal/common_nn.py:3072]: Is it important to still keep this around? Could we remove references to criterion_tests instead of having the empty list here?

**Review by janeyx99 — COMMENTED** [review]
Looks good overall--do you have an estimate of how much longer the new tests will take? I think this PR increases test coverage, and it'd be good to know the TTS add

File: torch/testing/_internal/common_nn.py (hunk 5 above)
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_nn.py:3072]: There are [a couple more tests being added to this list below](https://github.com/pytorch/pytorch/blob/b4defcb742e229e6343f3e21bab8b58d2ac1d8e4/torch/testing/_internal/common_nn.py#L3112-L3192) that weren't migrated in this PR

File: torch/testing/_internal/common_modules.py (hunk 3 above)
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_modules.py:2208]: oh good catch, thanks!

File: torch/testing/_internal/common_modules.py [hunk 6]
```diff
@@ -349,6 +420,105 @@ def module_inputs_torch_nn_GaussianNLLLoss(module_info, device, dtype, requires_
...
+    make_target = partial(make_tensor, device=device, dtype=dtype, requires_grad=False)
+
+    cases: List[Tuple[str, dict]] = [
+        ('', {}),
+        ('reduction_sum', {'reduction': 'sum'}),
+        ('reduction_mean', {'reduction': 'mean'}),
+        ('reduction_none', {'reduction': 'none'}),
+        ('full', {'full': True}),
+        ('no_log_input', {'log_input': False}),
+        ('full_no_log_input', {'full': True, 'log_input': False}),
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_modules.py:434]: This is the default that corresponds to the ('', {}) on line 428

File: torch/testing/_internal/common_modules.py [hunk 7]
```diff
@@ -315,12 +356,42 @@ def reference_fn(m, p, i, t, constructor_kwargs=constructor_kwargs):
 
         module_inputs.append(
             ModuleInput(constructor_input=FunctionInput(**constructor_kwargs),
-                        forward_input=FunctionInput(make_input((15, 10)),
-                                                    torch.empty(15, device=device).uniform_().mul(10).floor().long()),
+                        forward_input=FunctionInput(make_input((15, 4)),
+                                                    torch.empty(15, device=device).uniform_().mul(4).floor().long()),
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_modules.py:360]: Wanted to decrease the test time since there are quite a lot of sample inputs, I don't think 10 is a special number here so any other number of classes is probably also good :)

File: test/test_modules.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/test_modules.py:307]: good point

File: torch/testing/_internal/common_nn.py [hunk 8]
```diff
@@ -3052,16 +3043,16 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
         losses.append(-alpha[-2:].sum().log()[None])
     output = torch.cat(losses, 0)
     if reduction == 'mean':
-        return (output / target_lengths.to(dtype=output.dtype, device=output.device)).mean()
+        output = (output / target_lengths.to(dtype=output.dtype, device=output.device)).mean()
     elif reduction == 'sum':
-        return output.sum()
+        output = output.sum()
     output = output.to(dt)
     return output
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_nn.py:3050]: Hm the problem here is that on this line the dtype of `log_probs` is promoted to `float64`, so the `output.to(dt) at the end is necessary for all `reduction` types so that `float32` input --> `float32` output

**Comment by mikaylagawarecki** [comment]
@janeyx99 Updated the PR description with the change in test time

File: torch/testing/_internal/common_nn.py (hunk 5 above)
Comment (janeyx99) [reply torch/testing/_internal/common_nn.py:3072]: ah i see!

**Review by janeyx99 — APPROVED** [review]
LGTM!

**Comment by janeyx99** [comment]
it’s an OOM for MSELoss—is this expected? if so, we should just add one of those memory decorators 

https://hud.pytorch.org/pytorch/pytorch/commit/f138b08d2e9c8d676f2a404e97d773f42132b0c7

**Comment by mikaylagawarecki** [comment]
ciflow/slow seems to be passing now

Removed the extremely large ModuleInput with shape (5, 68, 64, 64) in MSELoss for mpw no other moduleinputs have such a big in[it tensor, going to look into where this large input was first added (it seems legacy) and might add it back in a follow up
